File: android/app/src/test/java/eu/euroswarms/surgeon/data/StoreTest.kt
package eu.euroswarms.surgeon.data

import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import java.io.File
import java.nio.file.Files
import java.util.UUID

class StoreTest {

    private lateinit var dir: File

    @Before
    fun setUp() {
        dir = Files.createTempDirectory("store-test").toFile()
    }

    private fun draft(status: DraftStatus = DraftStatus.DRAFT, createdAt: Long = System.currentTimeMillis()) =
        DraftPr(
            id = UUID.randomUUID().toString(),
            repoFullName = "up/repo",
            forkFullName = "me/repo",
            baseBranch = "main",
            headBranch = "surgeon/issue-1-x",
            issueNumber = 1,
            issueTitle = "t",
            issueUrl = "u",
            isPrivateRepo = false,
            status = status,
            createdAt = createdAt,
        )

    @Test
    fun processedMemoryPersistsAcrossInstances() = runBlocking {
        val store = Store(dir)
        assertFalse(store.isProcessed("up/repo", 7))
        store.markProcessed("up/repo", 7)
        assertTrue(store.isProcessed("up/repo", 7))
        // A fresh Store over the same directory must read it back.
        assertTrue(Store(dir).isProcessed("up/repo", 7))
    }

    @Test
    fun skippedMemoryIsSeparateAndClearable() = runBlocking {
        val store = Store(dir)
        store.markSkipped("up/repo", 9)
        assertTrue(store.isSkipped("up/repo", 9))
        assertFalse(store.isProcessed("up/repo", 9))
        assertEquals(1, store.skippedCount.value)

        // Persists across instances.
        assertTrue(Store(dir).isSkipped("up/repo", 9))

        // Clearing makes the issue eligible again.
        store.clearSkipped()
        assertFalse(store.isSkipped("up/repo", 9))
        assertEquals(0, store.skippedCount.value)
        assertFalse(Store(dir).isSkipped("up/repo", 9))
    }

    @Test
    fun draftUpsertAndStatusUpdate() = runBlocking {
        val store = Store(dir)
        val d = draft()
        store.upsertDraft(d)
        assertEquals(1, store.drafts.value.size)

        store.updateDraftStatus(d.id, DraftStatus.SUBMITTED)
        assertEquals(DraftStatus.SUBMITTED, store.drafts.value.first().status)

        // Persisted for a fresh instance.
        assertEquals(DraftStatus.SUBMITTED, Store(dir).drafts.value.first().status)
    }

    @Test
    fun draftsTodayExcludesDiscardedAndFailed() = runBlocking {
        val store = Store(dir)
        store.upsertDraft(draft(DraftStatus.DRAFT))
        store.upsertDraft(draft(DraftStatus.SUBMITTED))
        store.upsertDraft(draft(DraftStatus.DISCARDED))
        store.upsertDraft(draft(DraftStatus.FAILED))
        assertEquals(2, store.draftsToday())
    }

    @Test
    fun draftsTodayExcludesYesterday() = runBlocking {
        val store = Store(dir)
        store.upsertDraft(draft(createdAt = System.currentTimeMillis() - 48 * 3600 * 1000))
        assertEquals(0, store.draftsToday())
    }
}

File: web/tests/store.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { Store, MemoryStorage } from '../js/store.js';
import { DraftStatus } from '../js/models.js';

function draft(overrides = {}) {
  return {
    id: crypto.randomUUID(),
    repoFullName: 'up/repo',
    forkFullName: 'me/repo',
    baseBranch: 'main',
    headBranch: 'surgeon/issue-1-x',
    issueNumber: 1,
    issueTitle: 't',
    issueUrl: 'u',
    isPrivateRepo: false,
    status: DraftStatus.DRAFT,
    createdAt: Date.now(),
    ...overrides,
  };
}

test('processed memory persists across instances', () => {
  const storage = new MemoryStorage();
  const store = new Store(storage);
  assert.equal(store.isProcessed('up/repo', 7), false);
  store.markProcessed('up/repo', 7);
  assert.equal(store.isProcessed('up/repo', 7), true);
  assert.equal(new Store(storage).isProcessed('up/repo', 7), true);
});

test('skipped memory is separate and clearable', () => {
  const storage = new MemoryStorage();
  const store = new Store(storage);
  store.markSkipped('up/repo', 9);
  assert.equal(store.isSkipped('up/repo', 9), true);
  assert.equal(store.isProcessed('up/repo', 9), false);
  assert.equal(store.skippedCount(), 1);
  assert.equal(new Store(storage).isSkipped('up/repo', 9), true);

  store.clearSkipped();
  assert.equal(store.isSkipped('up/repo', 9), false);
  assert.equal(store.skippedCount(), 0);
  assert.equal(new Store(storage).isSkipped('up/repo', 9), false);
});

test('draft upsert and status update persist', () => {
  const storage = new MemoryStorage();
  const store = new Store(storage);
  const d = draft();
  store.upsertDraft(d);
  assert.equal(store.getDrafts().length, 1);

  store.updateDraftStatus(d.id, DraftStatus.SUBMITTED);
  assert.equal(store.getDrafts()[0].status, DraftStatus.SUBMITTED);
  assert.equal(new Store(storage).getDrafts()[0].status, DraftStatus.SUBMITTED);
});

test('draftsToday excludes discarded, failed, and yesterday', () => {
  const store = new Store(new MemoryStorage());
  store.upsertDraft(draft({ status: DraftStatus.DRAFT }));
  store.upsertDraft(draft({ status: DraftStatus.SUBMITTED }));
  store.upsertDraft(draft({ status: DraftStatus.DISCARDED }));
  store.upsertDraft(draft({ status: DraftStatus.FAILED }));
  store.upsertDraft(draft({ createdAt: Date.now() - 48 * 3600 * 1000 }));
  assert.equal(store.draftsToday(), 2);
});

test('config round-trips and merges defaults', () => {
  const storage = new MemoryStorage();
  const store = new Store(storage);
  const cfg = store.getConfig();
  assert.equal(cfg.authorName, 'Ilum');
  assert.equal(cfg.authorEmail, 'Ilum@linux.org');
  assert.equal(cfg.ollamaModel, 'deepseek-v4-flash:cloud');
  assert.equal(cfg.repos.length, 7);

  store.saveConfig({ ...cfg, githubToken: 'tok' });
  assert.equal(new Store(storage).getConfig().githubToken, 'tok');
});

test('logs are capped', () => {
  const store = new Store(new MemoryStorage());
  for (let i = 0; i < 350; i++) store.log('INFO', `m${i}`);
  assert.equal(store.getLogs().length, 300);
  assert.equal(store.getLogs()[0].message, 'm349');
});
